src/common/transformations/src/transformations/smart_reshape/lstm_states_broadcast.cpp

```diff
@@ -0,0 +1,171 @@
+    for (auto& parameter : body->get_parameters()) {
+        auto pshape = parameter->get_partial_shape();
+        original_shapes[parameter.get()] = pshape;
+        if (pshape.rank().is_dynamic())
+            continue;
+        for (ov::Dimension& n : pshape) {
+            n = ov::Dimension::dynamic();
+            ov::DimensionTracker::set_label(n, label++);
+        }
+        parameter->set_partial_shape(pshape);
+    }
+
+    // propagate labels through TI body
+    body->validate_nodes_and_infer_types();
+    // if lstm first input has undefined rank or if tracked label is zero -- we failed to track batch dimension
+    // returning body to initial state
+    if (lstm_cell->get_input_partial_shape(0).rank().is_dynamic() ||
+        ov::DimensionTracker::get_label(lstm_cell->get_input_partial_shape(0)[0]) == 0) {
+        for (auto& item : original_shapes)
+            item.first->set_partial_shape(item.second);
+        body->validate_nodes_and_infer_types();
+        return nullptr;
+    }
+
+    // batch label was tracked -- finding parameter that delivered it
+    shared_ptr<Parameter> batch_delivering_parameter;
+    size_t index_of_batch_dim = 0;
+
+    size_t batch_label = ov::DimensionTracker::get_label(lstm_cell->get_input_partial_shape(0)[0]);
+    for (auto& parameter : body->get_parameters()) {
+        auto pshape = parameter->get_partial_shape();
+        if (pshape.rank().is_dynamic())
+            continue;
+        for (size_t i = 0; i < pshape.size(); ++i) {
+            if (ov::DimensionTracker::get_label(pshape[i]) == batch_label) {
+                batch_delivering_parameter = parameter;
+                index_of_batch_dim = i;
+                break;
+            }
+        }
+        if (index_of_batch_dim != 0 && batch_delivering_parameter != nullptr)
+            break;
+    }
+    for (auto& item : original_shapes)
+        item.first->set_partial_shape(item.second);
+    body->validate_nodes_and_infer_types();
+
+    if (batch_delivering_parameter == nullptr)
+        return nullptr;
+
+    const auto& batched_source = get_outer_input_of_ti_by_parameter(batch_delivering_parameter, ti);
+    const auto& batched_shape = make_shared<ShapeOf>(batched_source.get_source_output());
+    const auto& batch = make_shared<Gather>(batched_shape,
+                                            Constant::create(ov::element::i64, ov::Shape{1}, {index_of_batch_dim}),
+                                            Constant::create(ov::element::i64, ov::Shape{}, {0}));
+    return batch;
+}
+
+bool broadcast_state_by_batch(ov::Input<ov::Node> input, const shared_ptr<ov::Node>& batch_delivering_node) {
+    auto constant_state = dynamic_pointer_cast<Constant>(input.get_source_output().get_node_shared_ptr());
+    if (constant_state == nullptr)
+        return false;
+    const auto& constant_shape = constant_state->get_shape();
+    if (constant_shape[0] != 1)
```

Comment:
do we need to check rank here? maybe constant is scalar

Reply:
It couldn't. We believe network is validated before and it goes into LSTMCell, so there is no need to check.